src/backend/renderer/damage/shaper.rs

```diff
@@ -0,0 +1,407 @@
+    // computed for each tile individually by walking all the current damage rectangles.
+    #[inline(never)]
+    fn shape_damage_impl(
+        &mut self,
+        in_damage: &mut [Rectangle<i32, Physical>],
+        last_direction: Option<DamageSplitAxis>,
+        invert_direction: bool,
+    ) {
+        // Optimize small damage input.
+        if in_damage.is_empty() {
+            return;
+        } else if in_damage.len() == 1 {
+            self.out_damage.push(in_damage[0]);
+            return;
+        }
+
+        // Compute the effective damage bounding box and the maximum damaged area rectangle.
+
+        let (x_min, y_min, x_max, y_max, max_damage_area) =
+            in_damage.iter().fold((i32::MAX, i32::MAX, 0, 0, 0), |acc, rect| {
+                let area = rect.size.w * rect.size.h;
+                (
+                    acc.0.min(rect.loc.x),
+                    acc.1.min(rect.loc.y),
+                    acc.2.max(rect.loc.x + rect.size.w),
+                    acc.3.max(rect.loc.y + rect.size.h),
+                    acc.4.max(area),
+                )
+            });
+
+        let bbox_w = x_max - x_min;
+        let bbox_h = y_max - y_min;
+
+        let damage_bbox = Rectangle::<i32, Physical>::from_loc_and_size((x_min, y_min), (bbox_w, bbox_h));
+
+        // Damage the current bounding box when there's a damage rect covering near all the area.
+        if max_damage_area as f32 / (damage_bbox.size.w * damage_bbox.size.h) as f32
+            > MAX_DAMAGE_TO_DAMAGE_BBOX_RATIO
+        {
+            self.out_damage.push(damage_bbox);
+            return;
+        }
+
+        // Now we try to split bounding box to process non-overlapping damage rects separately.
+        //
+        // The whole approach is recursive and splits viewport if and only if we have a gap
+        // in the current segment or rectangles touch each other, since the renderer excludes
+        // borders.
+        //
+        // Examples:
+        //      [0, 3], [1, 2] [2, 3] [3, 4]
+        //      will have a split at ^
+        //      [0, 3], [1, 2] [2, 3] [6, 10]
+        //      will have a split at ^
+        //
+        //  Resulting in recursively trying to shape damage before and after
+        //  split point.
+
+        let mut direction = if bbox_w >= bbox_h {
+            DamageSplitAxis::X
+        } else {
+            DamageSplitAxis::Y
+        };
+
+        if invert_direction {
+            direction = direction.invert();
+        }
+
+        // The coordinate where the first rectangle ends and where the potential overlap
+        // will end.
+        let mut overlap_end = match direction {
+            DamageSplitAxis::X => {
+                if Some(direction) != last_direction {
+                    in_damage.sort_unstable_by(|lhs, rhs| {
+                        // Sort ascending by X and then descending by width, so the first
+                        // rectangle will overlap the most.
```

Comment:
But it it's [0, 1], [1, 10], [2, 9] according to this sort then the first rectangle isn't the one that overlaps the most?

If this is intended then comments here should be clarified as to what's going on.

Reply:
_ascending by X_ **then** descending by _width_ everything you showed is correct.